boofuzz/primitives/bit_field.py

```diff
@@ -106,34 +107,59 @@ def __init__(
                     self._fuzz_library.append(val)
             else:
                 # try only "smart" values.
-                self.add_integer_boundaries(0)
-                self.add_integer_boundaries(self.max_num // 2)
-                self.add_integer_boundaries(self.max_num // 3)
-                self.add_integer_boundaries(self.max_num // 4)
-                self.add_integer_boundaries(self.max_num // 8)
-                self.add_integer_boundaries(self.max_num // 16)
-                self.add_integer_boundaries(self.max_num // 32)
-                self.add_integer_boundaries(self.max_num)
+                self.add_integer_boundaries(0, 10)
+                self.add_integer_boundaries(self.max_num // 2, 10)
+                self.add_integer_boundaries(self.max_num // 3, 10)
+                self.add_integer_boundaries(self.max_num // 4, 10)
+                self.add_integer_boundaries(self.max_num // 8, 10)
+                self.add_integer_boundaries(self.max_num // 16, 10)
+                self.add_integer_boundaries(self.max_num // 32, 10)
+                self.add_integer_boundaries(self.max_num, 10)
+
+                power_of_two = math.log(self.max_num, 2)
+                bits = width if not self.max_num else int(math.floor(power_of_two))
+                tmp = 1 << bits
+                while bits != 0 and tmp != 0:
+                    power_of_two = math.log(self.max_num, 2)
+                    self.add_integer_boundaries(tmp, 1)
+                    if signed:
+                        self.add_integer_boundaries(-tmp, 1)
+                    # check if width is a power of 2. if it isn't, calculate and add numbers near the next
+                    # lower power of two
+                    power_of_two = math.log(width, 2)
+                    if power_of_two != math.floor(power_of_two) and power_of_two > 0:
+                        tmp = 1 << int(power_of_two)
+                        self.add_integer_boundaries(tmp, 1)
+                        if signed:
+                            self.add_integer_boundaries(-tmp, 1)
+                    bits = bits // 2
+                    tmp = 1 << bits
```

Comment:
I suspect this whole chunk could be cleaned up a bit -- perhaps more helpful variable names or a modification of the layout. If you have a go I can make a second pass.